// @defra/interactive-map/providers/maplibre/src/utils/highlightFeatures.js
const groupFeaturesBySource = (map, selectedFeatures) => {
  const featuresBySource = {}

  selectedFeatures?.forEach(({ featureId, layerId, idProperty, geometry }) => {
    const layer = map.getLayer(layerId)

    if (!layer) {
      return
    }

    const sourceId = layer.source

    if (!featuresBySource[sourceId]) {
      featuresBySource[sourceId] = {
        ids: new Set(),
        fillIds: new Set(),
        idProperty,
        layerId,
        hasFillGeometry: false
      }
    }

    // Track whether any selected feature on this source is a polygon
    if (geometry && (geometry.type === 'Polygon' || geometry.type === 'MultiPolygon')) {
      featuresBySource[sourceId].hasFillGeometry = true
      featuresBySource[sourceId].fillIds.add(featureId)
    }

    featuresBySource[sourceId].ids.add(featureId)
  })

  return featuresBySource
}

const cleanupStaleSources = (map, previousSources, currentSources) => {
  previousSources.forEach(src => {
    if (!currentSources.has(src)) {
      const base = `highlight-${src}`
      const layers = [`${base}-fill`, `${base}-line`, `${base}-symbol`]
      layers.forEach(id => {
        if (map.getLayer(id)) {
          map.setFilter(id, ['==', 'id', ''])
        }
      })
    }
  })
}

const applyHighlightLayer = (map, id, type, sourceId, srcLayer, paint, filter) => {
  if (!map.getLayer(id)) {
    map.addLayer({
      id,
      type,
      source: sourceId,
      ...(srcLayer && { 'source-layer': srcLayer }),
      paint
    })
  }
  Object.entries(paint).forEach(([prop, value]) => {
    map.setPaintProperty(id, prop, value)
  })
  map.setFilter(id, filter)
  map.moveLayer(id)
}

const applySymbolHighlightLayer = (map, id, sourceId, srcLayer, originalLayerId, imageId, filter) => {
  if (!map.getLayer(id)) {
    map.addLayer({
      id,
      type: 'symbol',
      source: sourceId,
      ...(srcLayer && { 'source-layer': srcLayer }),
      layout: {
        'icon-image': imageId,
        'icon-anchor': map.getLayoutProperty(originalLayerId, 'icon-anchor') ?? 'center',
        'icon-allow-overlap': true
      }
    })
  }
  map.setLayoutProperty(id, 'icon-image', imageId)
  map.setFilter(id, filter)
  map.moveLayer(id)
}

const calculateBounds = (LngLatBounds, renderedFeatures) => {
  if (!renderedFeatures.length) {
    return null
  }

  const bounds = new LngLatBounds()

  renderedFeatures.forEach(f => {
    const add = (c) => typeof c[0] === 'number' ? bounds.extend(c) : c.forEach(add)
    add(f.geometry.coordinates)
  })

  return [bounds.getWest(), bounds.getSouth(), bounds.getEast(), bounds.getNorth()]
}

const getSelectedImageId = (map, imageId) => map._symbolImageMap?.[imageId] ?? null

/**
 * Update highlighted features using pure filters.
 * Supports fill, line and symbol geometry, multi-source, cleanup, and bounds.
 */
export function updateHighlightedFeatures ({ LngLatBounds, map, selectedFeatures, stylesMap }) {
  if (!map) {
    return null
  }

  const featuresBySource = groupFeaturesBySource(map, selectedFeatures)
  const renderedFeatures = []

  const currentSources = new Set(Object.keys(featuresBySource))
  const previousSources = map._highlightedSources || new Set()

  cleanupStaleSources(map, previousSources, currentSources)
  map._highlightedSources = currentSources

  // Apply highlights for current sources
  currentSources.forEach(sourceId => {
    const { ids, fillIds, idProperty, layerId, hasFillGeometry } = featuresBySource[sourceId]
    const baseLayer = map.getLayer(layerId)
    const srcLayer = baseLayer.sourceLayer

    // Use the actual feature geometry to determine highlight type
    const geom = hasFillGeometry ? 'fill' : baseLayer.type
    const base = `highlight-${sourceId}`

    const idExpression = idProperty ? ['get', idProperty] : ['id']
    const filter = ['in', idExpression, ['literal', [...ids]]]

    if (geom === 'fill') {
      const { stroke, strokeWidth, fill } = stylesMap[layerId]
      const fillFilter = ['in', idExpression, ['literal', [...fillIds]]]
      const linePaint = { 'line-color': stroke, 'line-width': strokeWidth }
      // Only apply fill highlight to polygon features, not to any co-selected line features
      applyHighlightLayer(map, `${base}-fill`, 'fill', sourceId, srcLayer, { 'fill-color': fill }, fillFilter)
      applyHighlightLayer(map, `${base}-line`, 'line', sourceId, srcLayer, linePaint, filter)
    }

    if (geom === 'line') {
      const { stroke, strokeWidth } = stylesMap[layerId]
      const linePaint = { 'line-color': stroke, 'line-width': strokeWidth }
      // Clear any fill highlight from a previous polygon selection on the same source
      if (map.getLayer(`${base}-fill`)) {
        map.setFilter(`${base}-fill`, ['==', 'id', ''])
      }
      applyHighlightLayer(map, `${base}-line`, 'line', sourceId, srcLayer, linePaint, filter)
    }

    if (geom === 'symbol') {
      const imageId = map.getLayoutProperty(layerId, 'icon-image')
      const selectedImageId = getSelectedImageId(map, imageId)
      if (selectedImageId) {
        applySymbolHighlightLayer(map, `${base}-symbol`, sourceId, srcLayer, layerId, selectedImageId, filter)
      }
    }

    // Bounds only from rendered tiles
    renderedFeatures.push(
      ...map
        .queryRenderedFeatures({ layers: [layerId] })
        .filter(f => ids.has(idProperty ? f.properties?.[idProperty] : f.id))
    )
  })

  return calculateBounds(LngLatBounds, renderedFeatures)
}
