// @defra/interactive-map/src/config/appConfig.js
import { KeyboardHelp } from '../App/components/KeyboardHelp/KeyboardHelp.jsx'

const keyboardBasePanelSlots = {
  slot: 'middle',
  open: false,
  dismissible: true,
  modal: true
}

const buttonSlots = {
  slot: 'right-top',
  showLabel: false
}

const exitButtonSlots = {
  slot: 'top-left',
  showLabel: false
}

// Default app buttons, panels and icons
export const defaultAppConfig = {
  buttons: [{
    id: 'exit',
    label: 'Exit',
    iconId: 'close',
    onClick: (_e, { services }) => services.closeApp(),
    excludeWhen: ({ appConfig, appState }) => !appConfig.hasExitButton || !appState.isFullscreen,
    mobile: exitButtonSlots,
    tablet: exitButtonSlots,
    desktop: exitButtonSlots
  }, {
    id: 'fullscreen',
    label: () => `${document.fullscreenElement ? 'Exit' : 'Enter'} fullscreen`,
    iconId: () => document.fullscreenElement ? 'minimise' : 'maximise',
    onClick: (_e, { appState }) => {
      const container = appState.layoutRefs.appContainerRef.current
      document.fullscreenElement ? document.exitFullscreen() : container.requestFullscreen()
    },
    excludeWhen: ({ appState, appConfig }) => !appConfig.enableFullscreen || appState.isFullscreen,
    mobile: buttonSlots,
    tablet: buttonSlots,
    desktop: buttonSlots
  }, {
    id: 'zoomIn',
    group: { name: 'zoom', label: 'Zoom controls', order: 0 },
    label: 'Zoom in',
    iconId: 'plus',
    onClick: (_e, { mapProvider, appConfig }) => mapProvider.zoomIn(appConfig.zoomDelta),
    excludeWhen: ({ appState, appConfig }) => !appConfig.enableZoomControls || appState.interfaceType === 'touch',
    enableWhen: ({ mapState }) => !mapState.isAtMaxZoom,
    mobile: buttonSlots,
    tablet: buttonSlots,
    desktop: buttonSlots
  }, {
    id: 'zoomOut',
    group: { name: 'zoom', label: 'Zoom controls', order: 0 },
    label: 'Zoom out',
    iconId: 'minus',
    onClick: (_e, { mapProvider, appConfig }) => mapProvider.zoomOut(appConfig.zoomDelta),
    excludeWhen: ({ appState, appConfig }) => !appConfig.enableZoomControls || appState.interfaceType === 'touch',
    enableWhen: ({ mapState }) => !mapState.isAtMinZoom,
    mobile: buttonSlots,
    tablet: buttonSlots,
    desktop: buttonSlots
  }],

  panels: [{
    id: 'keyboardHelp',
    label: 'Keyboard shortcuts',
    mobile: {
      ...keyboardBasePanelSlots
    },
    tablet: {
      ...keyboardBasePanelSlots,
      width: '500px'
    },
    desktop: {
      ...keyboardBasePanelSlots,
      width: '500px'
    },
    render: () => <KeyboardHelp />
  }],

  icons: [{
    id: 'maximise',
    svgContent: '<path d="M15 3h6v6"/><path d="m21 3-7 7"/><path d="m3 21 7-7"/><path d="M9 21H3v-6"/>'
  }, {
    id: 'minimise',
    svgContent: '<path d="m14 10 7-7"/><path d="M20 10h-6V4"/><path d="m3 21 7-7"/><path d="M4 14h6v6"/>'
  }, {
    id: 'plus',
    svgContent: '<path d="M5 12h14"/><path d="M12 5v14"/>'
  }, {
    id: 'minus',
    svgContent: '<path d="M5 12h14"/>'
  }, {
    id: 'chevron',
    svgContent: '<path d="m6 9 6 6 6-6"/>'
  }]
}

// Used by addButton
const defaultButtonSlots = {
  slot: 'right-top',
  showLabel: true
}

export const defaultButtonConfig = {
  label: 'Button',
  mobile: defaultButtonSlots,
  tablet: defaultButtonSlots,
  desktop: defaultButtonSlots
}

// Used by addPanel
export const defaultPanelConfig = {
  label: 'Panel',
  focus: true,
  mobile: {
    slot: 'drawer',
    open: true,
    dismissible: true,
    modal: false,
    showLabel: true
  },
  tablet: {
    slot: 'left-top',
    open: true,
    dismissible: true,
    modal: false,
    showLabel: true
  },
  desktop: {
    slot: 'left-top',
    open: true,
    dismissible: true,
    modal: false,
    showLabel: true
  },
  render: null,
  html: null
}

// Used by addControl
export const defaultControlConfig = {
  label: 'Control',
  mobile: {
    slot: 'drawer'
  },
  tablet: {
    slot: 'top-left'
  },
  desktop: {
    slot: 'top-left'
  }
}

export const scaleFactor = {
  small: 1,
  medium: 1.5,
  large: 2
}
