// @defra/interactive-map/src/App/components/MapButton/MapButton.jsx
// components/MapButton.jsx
import React, { useState, useRef } from 'react'
import { stringToKebab } from '../../../utils/stringToKebab'
import { Tooltip } from '../Tooltip/Tooltip'
import { Icon } from '../Icon/Icon'
import { SlotRenderer } from '../../renderer/SlotRenderer'
import { PopupMenu } from '../PopupMenu/PopupMenu'
import { useConfig } from '../../store/configContext'
import { useApp } from '../../store/appContext'

/**
 * Builds CSS class names for the map button element.
 * @param {string} buttonId - Unique identifier for the button (kebab-cased)
 * @param {string} variant - Visual variant of the button (e.g., 'primary', 'secondary')
 * @param {boolean} showLabel - Whether the button label is displayed
 * @returns {string} Space-separated CSS class names
 */
const buildButtonClassNames = (buttonId, variant, showLabel) => [
  'im-c-map-button',
  buttonId && `im-c-map-button--${stringToKebab(buttonId)}`,
  variant && `im-c-map-button--${variant}`,
  showLabel && 'im-c-map-button--with-label'
].filter(Boolean).join(' ')

/**
 * Builds CSS class names for the wrapper div that contains the button.
 * @param {string} buttonId - Unique identifier for the button
 * @param {boolean} showLabel - Whether the button label is displayed
 * @returns {string} Space-separated CSS class names for the wrapper
 */
const buildWrapperClassNames = (buttonId, showLabel) => [
  'im-c-button-wrapper',
  buttonId && `im-c-button-wrapper--${stringToKebab(buttonId)}`,
  showLabel && 'im-c-button-wrapper--wide'
].filter(Boolean).join(' ')

/**
 * Handles spacebar key presses on anchor links by triggering a click event.
 * This ensures anchor buttons behave like standard buttons when spacebar is pressed.
 * @param {KeyboardEvent} e - The keyboard event
 */
const handleKeyUp = (e) => {
  if (e.key === ' ' || e.key === 'Spacebar') {
    e.preventDefault()
    e.currentTarget.click()
  }
}

const captureMenuRect = (buttonRefs, buttonId, setMenuRect) => {
  const btn = buttonRefs.current[buttonId]
  if (!btn) {
    return
  }
  setMenuRect(btn.getBoundingClientRect().toJSON())
}

/**
 * Returns a keyup handler for buttons that control a popup menu.
 * ArrowDown opens the menu at the first item; ArrowUp opens at the last.
 * @param {boolean} hasMenu - Whether the button has a popup menu
 * @param {Object} buttonRefs - React ref map of button elements
 * @param {string} buttonId - Unique button identifier
 * @param {Function} setMenuStartPos - State setter for menu start position
 * @param {Function} setMenuRect - State setter for button bounding rect
 * @param {Function} setIsPopupOpen - State setter for popup open state
 * @returns {Function} Keyboard event handler
 */
const makePopupKeyUpHandler = (hasMenu, buttonRefs, buttonId, setMenuStartPos, setMenuRect, setIsPopupOpen) => (e) => {
  if (hasMenu && ['ArrowDown', 'ArrowUp'].includes(e.key)) {
    e.preventDefault()
    setMenuStartPos(e.key === 'ArrowUp' ? 'last' : 'first')
    captureMenuRect(buttonRefs, buttonId, setMenuRect)
    setIsPopupOpen(true)
  }
}

const getButtonSlot = (panelId, buttonId) =>
  panelId ? `${stringToKebab(buttonId)}-button` : undefined

/**
 * Determines the controlled element (panel or popup menu) for ARIA attributes.
 * @param {Object} options - Configuration options
 * @param {string} options.idPrefix - Prefix for generated IDs
 * @param {string} options.panelId - ID of the controlled panel (if applicable)
 * @param {string} options.buttonId - Unique button identifier
 * @param {boolean} options.hasMenu - Whether the button has a popup menu
 * @returns {Object|null} Object with id and type ('panel' or 'popup'), or null if no controlled element
 */
const getControlledElement = ({ idPrefix, panelId, buttonId, hasMenu }) => {
  if (panelId) {
    return { id: `${idPrefix}-panel-${stringToKebab(panelId)}`, type: 'panel' }
  }
  if (hasMenu) {
    return { id: `${idPrefix}-popup-${stringToKebab(buttonId)}`, type: 'popup' }
  }
  return null
}

/**
 * Builds the complete props object for a button or anchor element.
 * Handles ARIA attributes, event handlers, refs, and conditional styling.
 * @param {Object} options - Configuration options
 * @param {string} options.appId - Application identifier
 * @param {string} options.buttonId - Unique button identifier
 * @param {string} options.className - CSS classes for the button
 * @param {Function} options.onClick - Click event handler
 * @param {Function} options.onKeyUp - Key up event handler
 * @param {Object} options.buttonRefs - React ref object for storing button references
 * @param {boolean} options.isDisabled - Whether the button is disabled
 * @param {boolean} options.isPressed - Whether the button is in pressed state
 * @param {boolean} options.isExpanded - Whether content controlled by button is expanded
 * @param {boolean} options.isPanelOpen - Whether the controlled panel is open
 * @param {boolean} options.isPopupOpen - Whether the popup menu is open
 * @param {Object|null} options.controlledElement - The element controlled by this button
 * @param {string} options.href - URL for anchor element (if provided, renders as <a> instead of <button>)
 * @returns {Object} Props object suitable for button or anchor element
 */
const buildButtonProps = ({
  appId,
  buttonId,
  className,
  onClick,
  onKeyUp,
  buttonRefs,
  isDisabled,
  isPressed,
  isExpanded,
  isPanelOpen,
  isPopupOpen,
  controlledElement,
  href
}) => {
  let ariaExpanded
  if (controlledElement?.type === 'panel') {
    ariaExpanded = String(isPanelOpen)
  } else if (controlledElement?.type === 'popup') {
    ariaExpanded = isPopupOpen
  } else if (typeof isExpanded === 'boolean') {
    ariaExpanded = isExpanded
  } else {
    // No action
  }

  return {
    id: `${appId}-${stringToKebab(buttonId)}`,
    className,
    onClick,
    onKeyUp,
    ref: (el) => {
      if (buttonRefs.current && buttonId) {
        buttonRefs.current[buttonId] = el
      }
    },
    'aria-disabled': isDisabled || undefined,
    'aria-expanded': ariaExpanded,
    'aria-pressed': typeof isPressed === 'boolean' ? isPressed : undefined,
    'aria-controls': controlledElement?.id,
    'aria-haspopup': controlledElement?.type === 'popup' || undefined,
    ...(href
      ? { href, target: '_blank', onKeyUp: handleKeyUp, role: 'button' }
      : { type: 'button' })
  }
}

/**
 * MapButton component - A versatile button for map applications.
 * Supports icons, labels, popups, panels, tooltips, and button grouping.
 * Renders as either a <button> or <a> element depending on props.
 * @component
 * @param {Object} props - Component props
 * @param {string} props.buttonId - Unique identifier for the button
 * @param {string} [props.iconId] - Icon identifier to display
 * @param {string} [props.iconSvgContent] - SVG content as an alternative to iconId
 * @param {string} props.label - Button label/tooltip text
 * @param {boolean} [props.showLabel=false] - Whether to display the label visually
 * @param {boolean} [props.isDisabled=false] - Whether the button is disabled
 * @param {boolean} [props.isPressed] - Whether the button is in pressed state (aria-pressed)
 * @param {boolean} [props.isExpanded] - Whether content controlled by the button is expanded
 * @param {boolean} [props.isHidden=false] - Whether to hide the button (CSS display: none)
 * @param {boolean} [props.isPanelOpen=false] - Whether the controlled panel is open
 * @param {string} [props.variant] - CSS variant class for styling (e.g., 'primary')
 * @param {Function} [props.onClick] - Custom click handler
 * @param {string} [props.panelId] - ID of the panel controlled by this button
 * @param {Array<Object>} [props.menuItems] - Array of items for popup menu
 * @param {string} [props.idPrefix=''] - Prefix for generated panel/popup IDs
 * @param {string} [props.href] - URL for anchor element; if provided, renders as <a> instead of <button>
 * @returns {JSX.Element} The rendered button component
 */
export const MapButton = ({
  buttonId,
  iconId,
  iconSvgContent,
  label,
  showLabel,
  isDisabled,
  isPressed,
  isExpanded,
  isHidden,
  isPanelOpen,
  variant,
  onClick,
  panelId,
  menuItems,
  idPrefix,
  href
}) => {
  const { id: appId } = useConfig()
  const { buttonRefs } = useApp()
  const [isPopupOpen, setIsPopupOpen] = useState(false)
  const [menuStartPos, setMenuStartPos] = useState(null)
  const [menuRect, setMenuRect] = useState(null)
  const menuRef = useRef(null)

  const Element = href ? 'a' : 'button'
  const hasMenu = menuItems?.length >= 1
  const showIcon = iconId || iconSvgContent || hasMenu
  const buttonSlot = getButtonSlot(panelId, buttonId)
  const controlledElement = getControlledElement({ idPrefix, panelId, buttonId, hasMenu })

  /**
   * Handles button click events.
   * Toggles popup menu visibility if the button controls a popup.
   * Calls the custom onClick handler if provided.
   * @param {React.MouseEvent} e - The click event
   */
  const handleButtonClick = (e) => {
    if (isDisabled) {
      return
    }
    if (controlledElement?.type === 'popup') {
      const isKeyboard = e.nativeEvent.pointerType === ''
      /* istanbul ignore next as pointerType can't be tested in jest */
      setMenuStartPos(isKeyboard ? 'first' : null)
      if (!isPopupOpen) {
        captureMenuRect(buttonRefs, buttonId, setMenuRect)
      }
      setIsPopupOpen((prev) => !prev)
    }
    if (onClick) {
      onClick(e)
    }
  }

  const handleButtonKeyUp = makePopupKeyUpHandler(hasMenu, buttonRefs, buttonId, setMenuStartPos, setMenuRect, setIsPopupOpen)

  const buttonProps = buildButtonProps({
    appId,
    buttonId,
    className: buildButtonClassNames(buttonId, variant, showLabel),
    onClick: handleButtonClick,
    onKeyUp: handleButtonKeyUp,
    buttonRefs,
    isDisabled,
    isPressed,
    isExpanded,
    isPanelOpen,
    isPopupOpen,
    controlledElement,
    href
  })

  const buttonEl = (
    <Element {...buttonProps}>
      {showIcon && <Icon id={iconId} svgContent={iconSvgContent} isMenu={hasMenu} />}
      {showLabel && <span>{label}</span>}
    </Element>
  )

  return (
    <div
      className={buildWrapperClassNames(buttonId, showLabel)}
      data-button-slot={buttonSlot}
      style={isHidden ? { display: 'none' } : undefined}
    >
      {showLabel ? buttonEl : <Tooltip content={label}>{buttonEl}</Tooltip>}
      {buttonSlot && <SlotRenderer slot={buttonSlot} />}
      {isPopupOpen && <PopupMenu popupMenuId={controlledElement.id} buttonId={buttonId} startPos={menuStartPos} menuRef={menuRef} items={menuItems} setIsOpen={setIsPopupOpen} buttonRect={menuRect} />}
    </div>
  )
}
