// @docusaurus/utils/lib/vcs/vcsHardcoded.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VcsHardcoded = exports.VCS_HARDCODED_UNTRACKED_FILE_PATH = exports.VCS_HARDCODED_LAST_UPDATE_INFO = exports.VCS_HARDCODED_CREATION_INFO = void 0;
exports.VCS_HARDCODED_CREATION_INFO = {
    timestamp: 1490997600000, // 1st Apr 2017
    author: 'Creator',
};
exports.VCS_HARDCODED_LAST_UPDATE_INFO = {
    timestamp: 1539502055000, // 14th Oct 2018
    author: 'Author',
};
exports.VCS_HARDCODED_UNTRACKED_FILE_PATH = `file/path/${Math.random()}.mdx`;
/**
 * This VCS implementation always returns hardcoded values for testing purposes.
 * It is also useful in dev environments where VCS info is not important.
 * Reading information from the VCS can be slow and is not always necessary.
 */
exports.VcsHardcoded = {
    initialize: () => {
        // Noop
    },
    getFileCreationInfo: async (filePath) => {
        if (filePath === exports.VCS_HARDCODED_UNTRACKED_FILE_PATH) {
            return null;
        }
        return exports.VCS_HARDCODED_CREATION_INFO;
    },
    getFileLastUpdateInfo: async (filePath) => {
        if (filePath === exports.VCS_HARDCODED_UNTRACKED_FILE_PATH) {
            return null;
        }
        return exports.VCS_HARDCODED_LAST_UPDATE_INFO;
    },
};
