// @defra/interactive-map/providers/maplibre/src/maplibreProvider.js
/**
 * @typedef {import('../../../src/types.js').MapProvider} MapProvider
 * @typedef {import('../../../src/types.js').MapProviderConfig} MapProviderConfig
 */

import { DEFAULTS, supportedShortcuts } from './defaults.js'
import { scaleFactor } from '../../../src/config/appConfig.js'
import { cleanCanvas, applyPreventDefaultFix } from './utils/maplibreFixes.js'
import { attachMapEvents } from './mapEvents.js'
import { attachAppEvents } from './appEvents.js'
import { getAreaDimensions, getCardinalMove, getBboxFromGeoJSON, isGeometryObscured, getResolution, getPaddedBounds } from './utils/spatial.js'
import { createMapLabelNavigator } from './utils/labels.js'
import { updateHighlightedFeatures } from './utils/highlightFeatures.js'
import { queryFeatures } from './utils/queryFeatures.js'
import { setupHoverCursor } from './utils/hoverCursor.js'
import { registerSymbols } from './utils/symbolImages.js'
import { registerPatterns } from './utils/patternImages.js'

/**
 * MapLibre GL JS implementation of the MapProvider interface.
 *
 * @implements {MapProvider}
 */
export default class MapLibreProvider {
  /**
   * @param {Object} options - Constructor options.
   * @param {any} options.mapFramework - The MapLibre GL JS module.
   * @param {MapProviderConfig} [options.mapProviderConfig={}] - Provider configuration.
   * @param {Object} options.events - Event name constants.
   * @param {Object} options.eventBus - Event emitter for publishing map events.
   */
  constructor ({ mapFramework, mapProviderConfig = {}, events, eventBus }) {
    this.maplibreModule = mapFramework
    this.events = events
    this.eventBus = eventBus
    this.capabilities = {
      supportedShortcuts,
      supportsMapSizes: true
    }
    // Spread all config properties onto the instance
    Object.assign(this, mapProviderConfig)
  }

  /**
   * Initialize the map.
   *
   * @param {Object} config - Map initialization configuration.
   * @returns {Promise<void>}
   */
  async initMap (config) {
    const { container, padding, mapStyle, mapSize, center, zoom, bounds, pixelRatio, ...initConfig } = config
    this.mapStyleId = mapStyle?.id
    this.mapSize = mapSize
    const { Map: MaplibreMap } = this.maplibreModule
    const { events, eventBus } = this

    const map = new MaplibreMap({
      ...initConfig,
      container,
      style: mapStyle?.url,
      pixelRatio,
      padding,
      center,
      zoom,
      fadeDuration: 0,
      attributionControl: false,
      dragRotate: false,
      doubleClickZoom: false
    })

    // Disable rotation
    map.touchZoomRotate.disableRotation()

    // map.showPadding = true
    this.map = map

    // Set padding before bounds
    this.map.setPadding(padding)

    // Set bounds after padding
    if (bounds) {
      map.fitBounds(bounds, { duration: 0 })
    }

    applyPreventDefaultFix(map)
    cleanCanvas(map)

    attachMapEvents({
      map,
      events,
      eventBus,
      getCenter: this.getCenter.bind(this),
      getZoom: this.getZoom.bind(this),
      getBounds: this.getBounds.bind(this),
      getResolution: this.getResolution.bind(this)
    })

    attachAppEvents({
      mapProvider: this,
      map,
      events,
      eventBus
    })

    // Add highlight layer after map load
    map.on('load', () => {
      this.labelNavigator = createMapLabelNavigator(map, mapStyle?.mapColorScheme, events, eventBus)
    })

    this.eventBus.emit(events.MAP_READY, {
      map: this.map,
      mapStyleId: this.mapStyleId,
      mapSize: this.mapSize,
      crs: this.crs
    })
  }

  /** Destroy the map and clean up resources. */
  destroyMap () {
    this.setHoverCursor([])
    this.mapEvents?.remove()
    this.appEvents?.remove()

    this.mapEvents = null
    this.appEvents = null

    this.map.remove()
  }

  /**
   * Set pointer cursor on the map canvas when hovering over any of the given layer IDs.
   * Call with an empty array to remove all hover cursor listeners.
   *
   * @param {string[]} layerIds
   */
  setHoverCursor (layerIds) {
    if (!this.map) {
      return
    }
    this._onHoverMove = setupHoverCursor(this.map, layerIds, this._onHoverMove)
  }

  // ==========================
  // Side-effects
  // ==========================

  /**
   * Set map view with optional center and zoom.
   *
   * @param {Object} options - View options.
   * @param {[number, number]} [options.center] - Center coordinates [lng, lat].
   * @param {number} [options.zoom] - Zoom level.
   */
  setView ({ center, zoom }) {
    this.map.flyTo({
      center: center || this.getCenter(),
      zoom: zoom || this.getZoom(),
      duration: DEFAULTS.animationDuration
    })
  }

  /**
   * Zoom in by delta.
   *
   * @param {number} zoomDelta - Amount to zoom in.
   */
  zoomIn (zoomDelta) {
    this.map.easeTo({
      zoom: this.getZoom() + zoomDelta,
      duration: DEFAULTS.animationDuration
    })
  }

  /**
   * Zoom out by delta.
   *
   * @param {number} zoomDelta - Amount to zoom out.
   */
  zoomOut (zoomDelta) {
    this.map.easeTo({
      zoom: this.getZoom() - zoomDelta,
      duration: DEFAULTS.animationDuration
    })
  }

  /**
   * Pan map by pixel offset [x, y]. Positive x pans right, positive y pans down.
   *
   * @param {[number, number]} offset - Pixel offset [x, y].
   */
  panBy (offset) {
    this.map.panBy(offset, { duration: DEFAULTS.animationDuration })
  }

  /**
   * Fit map view to the specified bounds or GeoJSON geometry.
   *
   * @param {[number, number, number, number] | object} bounds - Bounds as [west, south, east, north], or a GeoJSON Feature, FeatureCollection, or geometry. Bbox is computed from GeoJSON using @turf/bbox.
   */
  fitToBounds (bounds) {
    const bbox = Array.isArray(bounds) ? bounds : getBboxFromGeoJSON(bounds)
    this.map.fitBounds(bbox, { duration: DEFAULTS.animationDuration })
  }

  /**
   * Set map padding as pixel insets from the top, bottom, left and right edges of the map.
   *
   * @param {{ top?: number, bottom?: number, left?: number, right?: number }} padding - Padding in pixels.
   */
  setPadding (padding) {
    this.map.setPadding(padding)
  }

  // ==========================
  // Feature highlighting
  // ==========================

  /**
   * @experimental Update highlighted features on the map.
   *
   * @param {any[]} selectedFeatures - Features to highlight.
   * @param {any} stylesMap - Style configuration for highlighting.
   * @returns {any}
   */
  updateHighlightedFeatures (selectedFeatures, stylesMap) {
    const { LngLatBounds } = this.maplibreModule
    return updateHighlightedFeatures({ LngLatBounds, map: this.map, selectedFeatures, stylesMap })
  }

  // ==========================
  // Map label (keyboard-friendly)
  // ==========================

  /**
   * @experimental Highlight the next label in the specified direction for keyboard navigation.
   *
   * @param {string} direction - Direction to navigate (e.g., 'up', 'down', 'left', 'right').
   * @returns {any}
   */
  highlightNextLabel (direction) {
    return this.labelNavigator?.highlightNextLabel(direction) || null
  }

  /**
   * @experimental Highlight the label nearest to the map center.
   *
   * @returns {any}
   */
  highlightLabelAtCenter () {
    return this.labelNavigator?.highlightLabelAtCenter() || null
  }

  /**
   * @experimental Clear any highlighted label.
   */
  clearHighlightedLabel () {
    return this.labelNavigator?.clearHighlightedLabel() || null
  }

  // ==========================
  // Read-only getters
  // ==========================

  /**
   * Get current center coordinates [lng, lat].
   *
   * @returns {[number, number]}
   */
  getCenter () {
    const coord = this.map.getCenter()
    return [Number(coord.lng.toFixed(DEFAULTS.coordinatePrecision)), Number(coord.lat.toFixed(DEFAULTS.coordinatePrecision))]
  }

  /**
   * Get current zoom level.
   *
   * @returns {number}
   */
  getZoom () {
    return Number(this.map.getZoom().toFixed(DEFAULTS.coordinatePrecision))
  }

  /**
   * Get current bounds as [west, south, east, north].
   *
   * @returns {[number, number, number, number]}
   */
  getBounds () {
    return this.map.getBounds().toArray().flat(1)
  }

  /**
   * Query rendered features at a screen pixel position (x from left edge, y from top edge of viewport).
   *
   * @param {{ x: number, y: number }} point - Screen pixel position.
   * @param {Object} [options]
   * @param {number} [options.radius] - Pixel radius to expand the query area. Results sorted closest-first.
   * @returns {any[]}
   */
  getFeaturesAtPoint (point, options) {
    return queryFeatures(this.map, point, options)
  }

  /**
   * Rasterise and register symbol images for the given pre-resolved symbol configs.
   * Delegates to the shared symbol image utility so any plugin's MapLibre adapter can
   * register symbols without importing provider internals directly.
   *
   * The pixel ratio is computed as device pixel ratio × map size scale factor so symbols
   * are rasterised at the correct resolution for the current device DPI and map size.
   *
   * @param {Object[]} symbolConfigs - Flat list of datasets/merged-sublayers with a symbol config.
   *   Callers are responsible for sublayer merging before passing configs here.
   * @param {Object} mapStyle - Current map style config (provides id, selectedColor, haloColor)
   * @param {Object} symbolRegistry
   * @returns {Promise<void>}
   */
  async registerSymbols (symbolConfigs, mapStyle, symbolRegistry) {
    const pixelRatio = (this.map.getPixelRatio() || 1) * (scaleFactor[this.mapSize] || 1)
    return registerSymbols(this.map, symbolConfigs, mapStyle, symbolRegistry, pixelRatio)
  }

  /**
   * Rasterise and register pattern images for the given pre-resolved pattern configs.
   * Delegates to the shared pattern image utility so any plugin's MapLibre adapter can
   * register patterns without importing provider internals directly.
   *
   * @param {Object[]} patternConfigs - Flat list of datasets/merged-sublayers with a pattern config.
   *   Callers are responsible for sublayer merging before passing configs here.
   * @param {string} mapStyleId
   * @param {Object} patternRegistry
   * @returns {Promise<void>}
   */
  async registerPatterns (patternConfigs, mapStyleId, patternRegistry) {
    return registerPatterns(this.map, patternConfigs, mapStyleId, patternRegistry)
  }

  // ==========================
  // Spatial helpers
  // ==========================

  /**
   * Get the dimensions of the visible map area as a formatted string (e.g., '400m by 750m').
   *
   * @returns {string}
   */
  getAreaDimensions () {
    const { LngLatBounds } = this.maplibreModule
    return getAreaDimensions(getPaddedBounds(LngLatBounds, this.map)) // Use padded bounds
  }

  /**
   * Get cardinal direction and distance between two coordinates [lng, lat]. Returns a formatted string (e.g., 'north 400m' or 'south 400m, west 750m').
   *
   * @param {[number, number]} from - Start coordinates [lng, lat].
   * @param {[number, number]} to - End coordinates [lng, lat].
   * @returns {string}
   */
  getCardinalMove (from, to) {
    return getCardinalMove(from, to)
  }

  /**
   * Get map resolution in meters per pixel.
   *
   * @returns {number}
   */
  getResolution () {
    return getResolution(this.map.getCenter(), this.map.getZoom())
  }

  /**
   * Convert map coordinates [lng, lat] to screen pixel position (x from left edge, y from top edge of viewport).
   *
   * @param {[number, number]} coords - Map coordinates [lng, lat].
   * @returns {{ x: number, y: number }} Screen pixel position.
   */
  mapToScreen (coords) {
    return this.map.project(coords)
  }

  /**
   * Convert screen pixel position (x from left edge, y from top edge of viewport) to map coordinates [lng, lat].
   *
   * @param {{ x: number, y: number }} point - Screen pixel position.
   * @returns {[number, number]} Map coordinates [lng, lat].
   */
  screenToMap (point) {
    const { lng, lat } = this.map.unproject([point.x, point.y])
    return [lng, lat]
  }

  /**
   * Returns true if the geometry's screen bounding box overlaps the given panel rectangle.
   *
   * @param {object} geojson - GeoJSON Feature, FeatureCollection, or geometry.
   * @param {DOMRect} panelRect - Bounding rect of the panel element (viewport coordinates).
   * @returns {boolean}
   */
  isGeometryObscured (geojson, panelRect) {
    return isGeometryObscured(geojson, panelRect, this.map)
  }
}
